metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java

```diff
@@ -491,6 +536,13 @@ public void run() throws Exception {
                 return;
             }
 
+            if (!curLeaderAndEpoch.equals(leaderAndEpoch)) {
```

Comment:
Are any of the previous actions/checks in this method worth running if the epoch is stale? Why not make also add `MetadataChangeEvent` to the set of events that should be skipped in this situation? Might need a different name than  `MigrationWriteEvent`, but does the generalization make sense?

Reply:
Yea I wondered about this. In general, I'd like to avoid changing too much in MetadataChangeEvent.

Notice how MetadataChangeEvent doesn't use checkDriverState. This is because it needs to check both the driver state as well as the metadata log state. For the driver check this event can actually run in multiple states (since it's how we learn about brokers registering). It also has to check the migration state from the log to avoid trying to do dual writes before the migration is complete (the driver will see partial metadata get published for versions not including metadata transactions).